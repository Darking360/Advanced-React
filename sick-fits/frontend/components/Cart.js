import React from 'react';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';
import User from './User';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';

const LOCAL_STATE_QUERY = gql`
    query {
        cartOpen @client
    }
`;

const TOGGLE_CART_MUTATION = gql`
    mutation {
        toggleCart @client
    }
`;

const Cart = () => {
    return (
        <User>
            {({ data: { me } }) => {
                console.log(me);
                if (!me) return null;
                else return (
                    <Mutation mutation={TOGGLE_CART_MUTATION}>
                    {(toggleCart) => (
                        <Query query={LOCAL_STATE_QUERY}>
                            {({ data }) => (
                                <CartStyles open={data.cartOpen}>
                                    <header>
                                        <CloseButton title="close" onClick={toggleCart}>&times;</CloseButton>
                                        <Supreme>Your Cart</Supreme>
                                        <p>{me.name} have {me.cart.length} item{me.cart.length !== 1 ? 's' : null} in your cart</p>
                                    </header>
                                    <footer>
                                        <p>$10.10</p>
                                        <SickButton>Checkout</SickButton>
                                    </footer>
                                </CartStyles>
                            )}
                        </Query>
                    )}
                    </Mutation>
                )
            }}
        </User>
    );
}

export default Cart;
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION};