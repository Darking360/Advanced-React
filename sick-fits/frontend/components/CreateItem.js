import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import gql from 'graphql-tag';
import Error from './ErrorMessage';
import Router from 'next/router';

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
      title
    }
  }
`;

class CreateItem extends Component {

  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0,
  };

  handleChange = ({ target }) => {
    const { name, type, value } = target;
    const val = type === 'number' ? parseFloat(value) : value;
    this.setState({ [name]: val });
  }

  uploadFile = async (e) => {
    const files = e.target.files;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'sickfits');
    const response = await fetch(
      'https://api.cloudinary.com/v1_1/darking360/image/upload',
      {
        method: 'POST',
        body: data
      }
    );

    const file = await response.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    });

  }

  render() {
    const { title, price, description, image } = this.state;
    return(
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form data-test="form" onSubmit={async (e) => {
              e.preventDefault();
              const res = await createItem();
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id }
              });
            }}>
            <h2>Sell an Item</h2>
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">Title</label>
              <input
                type="file"
                id="file"
                name="file"
                placeholder="Upload an image"
                required
                onChange={this.uploadFile}
              />
              { image && <img src={image} width="200" alt="Image product preview"/> }
              <label htmlFor="title">Title</label>
              <input
                type="text"
                id="title"
                name="title"
                placeholder="Title"
                required
                value={title}
                onChange={this.handleChange}
              />
              <label htmlFor="price">Price</label>
              <input
                type="number"
                id="price"
                name="price"
                placeholder="Item's price"
                required
                value={price}
                onChange={this.handleChange}
              />
              <label htmlFor="description">Description</label>
              <textarea
                id="description"
                name="description"
                placeholder="Enter a description"
                required
                value={description}
                onChange={this.handleChange}
              />
              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }

}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
