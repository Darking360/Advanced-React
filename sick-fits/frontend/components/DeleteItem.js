import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from './Items';

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

class DeleteItem extends Component {

  handleDelete = (deleteItem) => {
    if (confirm('Are you sure you want to delete this?'))
      deleteItem()
      .catch((error) => alert(error.message));
  }

  update = (cache, payload) => {
    const { id } = this.props;
    // Manually update cache
    // Read the cache
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    data.items = data.items.filter((item) => item.id !== payload.data.deleteItem.id);
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
  }

  render() {
    const { id } = this.props;
    return(
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{ id }}
        update={this.update}
      >
        {(deleteItem, { error }) => (
          <button onClick={() => this.handleDelete(deleteItem)}>Delete</button>
        )}
      </Mutation>
    );
  }

}

export default DeleteItem;
