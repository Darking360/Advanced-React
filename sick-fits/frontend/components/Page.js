import React, { Component } from 'react';
import Header from './Header';
import Meta from './Meta';
import styled from 'styled-components';

const Button = styled.button`
  background: red;
  font-size: ${({huge}) => huge ? '100px' : '50px'};
`;

class Page extends Component {

  render() {
    const { children } = this.props;
    return(
      <div>
        <Meta />
        <Header />
        <Button huge>Click me</Button>
        { children }
      </div>
    )
  }
}

export default Page;
