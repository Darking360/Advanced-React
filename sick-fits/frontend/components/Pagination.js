import React, { Component } from 'react';
import PaginationStyles from './styles/PaginationStyles';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import Head from 'next/head';
import Link from 'next/link';
import { perPage } from '../config';

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

class Pagination extends Component {

  render() {
    const { page } = this.props;
    return(
      <Query query={PAGINATION_QUERY}>
        {({ loading, data, error }) => {
          if (loading) return <p>Loading...</p>
          else {
            const count = data.itemsConnection.aggregate.count;
            const pages = Math.ceil(count/perPage);
            return(
              <PaginationStyles data-test="pagination">
                <Head>
                  <title>Sick Fits! Page {page} of {pages}</title>
                </Head>
                <Link prefetch href={{ pathname: 'items', query: { page: page - 1 } }}>
                  <a className="prev" aria-disabled={page <= 1}>← Prev</a>
                </Link>
                <p>You are on page {page} of <span className="totalPages">{pages}</span></p>
                <p>{count} Items Total</p>
                <Link prefetch href={{ pathname: 'items', query: { page: page + 1 } }}>
                  <a className="next" aria-disabled={page >= pages}>Next →</a>
                </Link>
              </PaginationStyles>
            );
          }
        }}
      </Query>
    );
  }

}

export default Pagination;
export { PAGINATION_QUERY };
