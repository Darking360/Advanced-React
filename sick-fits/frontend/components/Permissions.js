import { Query, Mutation } from 'react-apollo';
import Error from './ErrorMessage';
import gql from 'graphql-tag';
import Table from './styles/Table';
import SickButton from './styles/SickButton';
import PropTypes from 'prop-types';

const possiblePermissions = [
    'ADMIN',
    'USER',
    'ITEMCREATE',
    'ITEMUPDATE',
    'ITEMDELETE',
    'PERMISSIONUPDATE',
  ];

const ALL_USERS_QUERY = gql`
    query {
        users {
            id
            name
            email
            permissions
        }
    }
`;

const UPDATE_PERMISSIONS_MUTATION = gql`
  mutation updatePermissions($permissions: [Permission], $userId: ID!) {
    updatePermissions(permissions: $permissions, userId: $userId) {
        id
        permissions
        name
        email
    }
  }
`;

const Permissions = (props) => (
    <Query query={ALL_USERS_QUERY}>
        {({ data, loading, error }) => (
            <div>
                <Error error={error} />
                <h2>Manage permissions</h2>
                <Table>
                    <thead>
                        <tr>
                            <th>Name</th>
                            <th>Email</th>
                            { possiblePermissions.map((item) => (<th key={item}>{item}</th>)) }
                            <th>ðŸ‘‡</th>
                        </tr>
                    </thead>
                    <tbody>
                        {data.users.map((user) => <UserPermissions key={user.id} user={user} />)}
                    </tbody>
                </Table>
            </div>
        )}
    </Query>
)

class UserPermissions extends React.Component {
    static propTypes = {
        user: PropTypes.shape({
            name: PropTypes.string,
            email: PropTypes.string,
            id: PropTypes.string,
            permissions:  PropTypes.array,
        }).isRequired,
    };
    state = {
        permissions: this.props.user.permissions,
    };
    handlePermissionsChanged = ({ target: { value, checked } }) => {
        const { permissions } = this.state;
        if (checked) this.setState({ permissions: [...permissions, value] });
        else this.setState({ permissions: permissions.filter((item) => item !== value) });
    }
    render(){
        const { user } = this.props;
        return(
            <Mutation
                mutation={UPDATE_PERMISSIONS_MUTATION}
                variables={{
                    permissions: this.state.permissions,
                    userId: this.props.user.id,
                }}
            >
                <tr>
                    <td>{user.name}</td>
                    <td>{user.email}</td>
                    { possiblePermissions.map((item) => {
                        return(
                            <td key={item}>
                                <label htmlFor={`${user.id}-permission-${item}`}>
                                    <input 
                                        id={`${user.id}-permission-${item}`}
                                        type="checkbox" 
                                        checked={this.state.permissions.includes(item)}
                                        value={item}
                                        onChange={this.handlePermissionsChanged} 
                                    />
                                </label>
                            </td>
                        );
                    }) }
                    <td>
                        <SickButton>UPDATE</SickButton>
                    </td>
                </tr>
            </Mutation>
        );
    }
}

export default Permissions;