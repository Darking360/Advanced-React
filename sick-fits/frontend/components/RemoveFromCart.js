import React from 'react';
import { Mutation } from 'react-apollo';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { CURRENT_USER_QUERY } from './User';

const BigButton = styled.button`
    font-size: 4rem;
    background: none;
    border: none;
    &:hover {
        color: ${({ theme }) => theme.red};
        cursor: pointer;

    }
`;

const REMOVE_FROM_CART_MUTATION = gql`
    mutation removeFromCart($id: ID!) {
        removeFromCart(id: $id) {
            id
        }
    }
`;

class RemoveFromCart extends React.Component {
    static propTypes = {
        id: PropTypes.string.isRequired,
    };

    // Gets called as soon as we get a response back from the server after the mutation
    update = (cache, payload) => {
        // Read the cache
        const data = cache.readQuery({ query: CURRENT_USER_QUERY });
        // Remove the item from the cart
        const cartItemId = payload.data.removeFromCart.id;
        data.me.cart = data.me.cart.filter((item) => item.id !== cartItemId);
        // Write it back to the cache
        cache.writeQuery({ query: CURRENT_USER_QUERY, data });
    }

    render() {
        const { id } = this.props;
        return (
            <Mutation
                mutation={REMOVE_FROM_CART_MUTATION}
                variables={{ id }}
                update={this.update}
                optimisticResponse={{
                    __typename: 'Mutation',
                    removeFromCart: { __typename: 'CartItem', id }
                }}
            >
                {(removeFromCart, { loading }) => (
                    <BigButton 
                        disabled={loading} 
                        title="Delete Item"
                        onClick={() => {
                            removeFromCart().catch((err) => alert(err.message));
                        }}
                    >
                        &times;
                    </BigButton>
                )}
            </Mutation>
        );
    }
}

export default RemoveFromCart;