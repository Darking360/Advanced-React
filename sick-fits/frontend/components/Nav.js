import Link from 'next/link';
import NavStyles from './styles/NavStyles';
import User from './User';
import Signout from './Signout';
import CartCount from './CartCount';
import { TOGGLE_CART_MUTATION } from './Cart';
import { Mutation } from 'react-apollo';

const Nav = props => (
  <User>
    {({ data: { me } }) => (
      <NavStyles>
        <Link href="/items">
          <a>Shop</a>
        </Link>
        { me && (
          <>
            <Link href="/sell">
              <a>Sell</a>
            </Link>
            <Link href="/orders">
              <a>Orders</a>
            </Link>
            <Link href="/me">
              <a>Account</a>
            </Link>
            <Signout />
            <Mutation mutation={TOGGLE_CART_MUTATION}>
              {(toggleCart) => (
                  <button onClick={toggleCart}>Cart <CartCount count={me.cart.reduce((tall, item) => tall + item.quantity, 0)} /></button>
              )}
            </Mutation>
          </>
        )}
        { !me && (
          <Link href="/signup">
            <a>Signin</a>
          </Link>
        )}
      </NavStyles>
    )}
  </User>
);

export default Nav;
