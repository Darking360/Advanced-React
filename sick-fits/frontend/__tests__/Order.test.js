import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import NProgress from 'nprogress';
import Router from 'next/router';
import Order, { SINGLE_ORDER_QUERY } from '../components/Order';
import { MockedProvider } from 'react-apollo/test-utils';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeOrderItem, fakeUser } from '../lib/testUtils';
import formatMoney from '../lib/formatMoney';

const items = [fakeOrderItem(), fakeOrderItem()];
const total = items.reduce((tall, node) => tall + node.quantity * node.price, 0);

const mocks = [
  {
    request: {
      query: SINGLE_ORDER_QUERY,
      variables: { id: 'abc123' },
    },
    result: {
      data: {
        order: {
          ...fakeOrderItem(),
          charge: 'xyz789',
          total, 
          items,
          user: fakeUser(),
          createdAt: new Date(),
        },
      },
    },
  },
]

describe('<Order />', () => {
  
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Order id='abc123' />
      </MockedProvider>
    );

    await wait();
    wrapper.update();

    const order = wrapper.find('div[data-test="order"]');
    expect(toJSON(order)).toMatchSnapshot();

  });

  it('renders correct items length', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Order id='abc123' />
      </MockedProvider>
    );

    await wait();
    wrapper.update();

    const orderItems = wrapper.find('.order-item');
    expect(toJSON(orderItems)).toHaveLength(2);
  });

  it('shows the correct charge, total, and items count', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Order id='abc123' />
      </MockedProvider>
    );

    await wait();
    wrapper.update();

    expect(wrapper.find('span[data-test="total"]').text()).toEqual(formatMoney(total));
    expect(wrapper.find('span[data-test="charge"]').text()).toEqual('xyz789');
    expect(wrapper.find('span[data-test="count"]').text()).toEqual(items.length.toString());
  });

});