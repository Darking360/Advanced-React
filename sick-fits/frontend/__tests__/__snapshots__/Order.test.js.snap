// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order /> renders and matches snapshot 1`] = `
<div
  className="OrderStyles-sc-4oqalm-0 camMij"
  data-test="order"
>
  <SideEffect(Head)>
    <Head />
  </SideEffect(Head)>
  <p>
    <span>
      Order ID
    </span>
    <span>
      9f8e79c4-0fa4-4505-a740-e67ee17558f5
    </span>
  </p>
  <p>
    <span>
      Charge
    </span>
    <span
      data-test="charge"
    >
      xyz789
    </span>
  </p>
  <p>
    <span>
      Date
    </span>
    <span>
      November 30, 2018 11:39 AM
    </span>
  </p>
  <p>
    <span>
      Order total
    </span>
    <span
      data-test="total"
    >
      $84.68
    </span>
  </p>
  <p>
    <span>
      Item Count
    </span>
    <span
      data-test="count"
    >
      2
    </span>
  </p>
  <div
    className="items"
  >
    <div
      className="order-item"
      key="9f8e79c4-0fa4-4505-a740-e67ee17558f5"
    >
      <img
        alt="soluta non omnis consequatur enim quia autem"
        src="reprehenderit.jpg"
      />
      <div
        className="item-details"
      >
        <h2>
          soluta non omnis consequatur enim quia autem
        </h2>
        <p>
          Quantity: 
          1
        </p>
        <p>
          Each: 
          $42.34
        </p>
        <p>
          Subtotal: 
          $42.34
        </p>
        <p>
          et modi tenetur amet modi reprehenderit omnis
        </p>
      </div>
    </div>
    <div
      className="order-item"
      key="9f8e79c4-0fa4-4505-a740-e67ee17558f5"
    >
      <img
        alt="quia sed exercitationem omnis laborum exercitationem est"
        src="sint.jpg"
      />
      <div
        className="item-details"
      >
        <h2>
          quia sed exercitationem omnis laborum exercitationem est
        </h2>
        <p>
          Quantity: 
          1
        </p>
        <p>
          Each: 
          $42.34
        </p>
        <p>
          Subtotal: 
          $42.34
        </p>
        <p>
          sapiente laudantium molestias assumenda quasi adipisci mollitia
        </p>
      </div>
    </div>
  </div>
</div>
`;
